## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [11]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [12]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [13]:
# Use the describe function to gather some basic statistics
df.describe()



client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [14]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.groupby(['category']).count()

first   last    job  phone  email  client_id  order_id  \
category                                                              
consumables  23538  23538  23538  23538  23538      23538     23538   
decor         7954   7954   7954   7954   7954       7954      7954   
electronics   2832   2832   2832   2832   2832       2832      2832   
furniture    11915  11915  11915  11915  11915      11915     11915   
software      8400   8400   8400   8400   8400       8400      8400   

             order_date  order_week  order_year  item_id  subcategory  \
category                                                                
consumables       23538       23538       23538    23538        23538   
decor              7954        7954        7954     7954         7954   
electronics        2832        2832        2832     2832         2832   
furniture         11915       11915       11915    11915        11915   
software           8400        8400        8400     8400         8400   

             unit_price  unit_cost  unit_weight    qty  line_number  
category                                                             
consumables       23538      23538        23538  23538        23538  
decor              7954       7954         7954   7954         7954  
electronics        2832       2832         2832   2832         2832  
furniture         11915      11915        11915  11915        11915  
software           8400       8400         8400   8400         8400

In [15]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
top_categories




category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [16]:

df[df['category'] == 'consumables'].groupby('subcategory')['subcategory'].count()

subcategory
bathroom supplies    6424
kitchen supplies     5746
misc                 2891
paper products       5874
pens                 2603
Name: subcategory, dtype: int64

In [17]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [18]:
# Store the client ids of those top 5 clients in a list.
top5_client_ids = df['client_id'].value_counts().head(5).index.tolist()

top5_client_ids

[33615, 66037, 46820, 38378, 24741]

In [19]:
# How many total units (the qty column) did the client with the most entries order order?
df[df['client_id'] == top5_client_ids[0]]['qty'].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [ ]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['qty'] * df['unit_price']  
df.head()



first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number  line_subtotal  shipping_price  total_price  line_cost  \
0                1      115164.00         5512.50    131839.08   85597.05   
1                0         523.95          312.90       914.26     629.79   
2                6         527.28          458.64      1077.12     765.18   
3                3        1056.18          356.70      1543.57    1077.35   
4                1        3902.00         6500.20     11364.40    8663.60   
...            ...            ...             ...          ...        ...   
54634            8        2743.29          519.75      3564.87    2222.55   
54635            0        9709.73         3849.30     14813.24   12095.92   
54636            6       31188.50        13832.00     49184.90   41054.25   
54637            0         165.76        14143.36     15632.71   14280.00   
54638            5        3103.31        14939.19     19711.43   16599.10   

       line_profit  
0         46242.03  
1           284.47  
2           311.94  
3           466.22  
4          2700.80  
...            ...  
54634      1342.32  
54635      2717.32  
54636      8130.65  
54637      1352.71  
54638      3112.33  

[54639 rows x 23 columns]

In [21]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound 
# for items 50 pounds or under.
def calculate_shipping_price(unit_weight, qty):
    total_weight = unit_weight * qty
    
    if total_weight > 50:
        price_per_pound = 7
    else:
        price_per_pound = 10
    
    shipping_price = total_weight * price_per_pound
    return shipping_price


df['shipping_price'] = df.apply(lambda row: calculate_shipping_price(row['unit_weight'], row['qty']), axis=1)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  \
0     1096.80     762.71         7.50  105            1      115164.00   
1       24.95      15.09         1.49   21            0         523.95   
2       13.52       7.86         1.68   39            6         527.28   
3       36.42      24.85         1.23   29            3        1056.18   
4      195.10     108.17        46.43   20            1        3902.00   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20

In [22]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = ((df['line_subtotal'] + df['shipping_price']) * 1.0925).round(2)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number  line_subtotal  shipping_price  \
0     762.71         7.50  105            1      115164.00         5512.50   
1      15.09         1.49   21            0         523.95          312.90   
2       7.86         1.68   39            6         527.28          458.64   
3      24.85         1.23   29            3        1056.18          356.70   
4     108.17        46.43   20            1        3902.00         6500.20   

   total_price  
0    131839.08  
1       914.26  
2      1077.12  
3      1543.57  
4     11364.40  

[5 rows x 21 columns]

In [23]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number  line_subtotal  shipping_price  \
0             7.50   105            1      115164.00         5512.50   
1             1.49    21            0         523.95          312.90   
2             1.68    39            6         527.28          458.64   
3             1.23    29            3        1056.18          356.70   
4            46.43    20            1        3902.00         6500.20   
...            ...   ...          ...            ...             ...   
54634         2.25    33            8        2743.29          519.75   
54635        11.70    47            0        9709.73         3849.30   
54636         4.16   475            6       31188.50        13832.00   
54637        18.04   112            0         165.76        14143.36   
54638         2.07  1031            5        3103.31        14939.19   

       total_price  line_cost  
0        131839.08   85597.05  
1           914.26     629.79  
2          1077.12     765.18  
3          1543.57    1077.35  
4         11364.40    8663.60  
...            ...        ...  
54634      3564.87    2222.55  
54635     14813.24   12095.92  
54636     49184.90   41054.25  
54637     15632

In [24]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number  line_subtotal  shipping_price  total_price  line_cost  \
0                1      115164.00         5512.50    131839.08   85597.05   
1                0         523.95          312.90       914.26     629.79   
2                6         527.28          458.64      1077.12     765.18   
3                3        1056.18          356.70      1543.57    1077.35   
4                1        3902.00         6500.20     11364.40    8663.60   
...            ...            ...             ...          ...        ...   
54634            8        2743.29          519.75      3564.87    2222.55   
54635            0        9709.73         3849.30     14813.24   12095.92   
54636            6       31188.50        13832.00     49184.90   41054.25   
54637            0         165.76        14143.36     15632.71   14280.00   
54638            5        3103.31        14939.19     19711.43   16599.10   

       line_profit  
0         46242.03  
1           284.47  
2           311.94  
3           466.22  
4          2700.80  
...            ...  
54634      1342.32  
54635      2717.32  
54636      8130.65  
54637      1352.71  
54638      3112.33  

[54639 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [25]:
# Check your work using the totals above
target_order_ids = [2742071, 2173913, 6128929]

# Calculate the total price for each target order ID
order_total = df[df['order_id'].isin(target_order_ids)].groupby('order_id')['total_price'].sum()

# Display the calculated total prices for comparison
print(order_total)


order_id
2173913    162388.71
2742071    152811.89
6128929    923441.25
Name: total_price, dtype: float64


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [45]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
filtered_df = df[df['client_id'].isin(top5_client_ids)]
total_spent = filtered_df.groupby('client_id')['total_price'].sum()
total_spent


client_id
24741    82268892.02
33615     8377308.52
38378    12906550.87
46820     9743794.36
66037    10259514.79
Name: total_price, dtype: float64

In [51]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
final_summary_df = filtered_df.groupby('client_id').agg(
    qty=('qty', 'sum'),
    shipping_price=('shipping_price', 'sum'),
    line_price=('total_price', 'sum'),
    line_cost=('line_cost', 'sum'),
    line_profit=('line_profit', 'sum')
).reset_index()
final_summary_df

client_id     qty  shipping_price   line_price    line_cost  line_profit
0      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
1      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
2      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
3      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
4      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81

In [52]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = ['shipping_price', 'line_price', 'line_cost', 'line_profit']

# Define a function that converts a dollar amount to millions.

def currency_format_millions(value):
    return value / 1_000_000

# Apply the currency_format_millions function to only the money columns. 

for col in money_columns:
    final_summary_df[col] = final_summary_df[col].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
final_summary_df = final_summary_df.rename(columns={
    'client_id': 'Client ID', 
    'qty': 'Units', 
    'shipping_price': 'Shipping (millions)', 
    'line_price': 'Total Revenue (millions)', 
    'line_profit': 'Total Profit (millions)'})

# Display the updated DataFrame
final_summary_df


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      24741  239862             5.126448                 82.268892   
1      33615   64313             1.828985                  8.377309   
2      38378   73667             3.429455                 12.906551   
3      46820   75768             1.601449                  9.743794   
4      66037   43018             1.395152                 10.259515   

   line_cost  Total Profit (millions)  
0  45.688900                36.579992  
1   6.175314                 2.201995  
2   9.634721                 3.271830  
3   7.007192                 2.736603  
4   7.004483                 3.255032

In [53]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
final_summary_df.sort_values(by="Total Profit (millions)", ascending=False).reset_index(drop=True)


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      24741  239862             5.126448                 82.268892   
1      38378   73667             3.429455                 12.906551   
2      66037   43018             1.395152                 10.259515   
3      46820   75768             1.601449                  9.743794   
4      33615   64313             1.828985                  8.377309   

   line_cost  Total Profit (millions)  
0  45.688900                36.579992  
1   9.634721                 3.271830  
2   7.004483                 3.255032  
3   7.007192                 2.736603  
4   6.175314                 2.201995

In [ ]:
'''Based on the analysis of the top 5 clients, Client 24741 stands out as the most significant 
contributor to the company's revenue and profit. This client generated $82.27 million in total 
revenue and $36.58 million in profit, far surpassing the performance of other top clients. 
The remaining four clients show more modest but still substantial contributions, with total 
revenues ranging from $8.38 million to $12.91 million and profits between $2.20 million and 
$3.27 million.'''